bindless resources demonstration

gpx1000 opened the pull request:
@wlohry thanks for the effort on this!

## Description

This sample demonstrates how to reduce CPU usage by offloading draw call generation and frustum culling to the GPU.

assets PR https://github.com/KhronosGroup/Vulkan-Samples-Assets/pull/13 required for this sample

## Tested Environments
Windows 10 / Ubuntu Linux
GeForce 3060 / GeForce 2080 / GeForce 1070
Vulkan version: 1.2.176

## General Checklist:

Please ensure the following points are checked:

- [X] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [X] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [X] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [X] I have commented any added functions (in line with Doxygen)
- [X] I have commented any code that could be hard to understand
- [X] My changes do not add any new compiler warnings
- [X] My changes do not add any new validation layer errors or warnings
- [X] I have used existing framework/helper functions where possible
- [X] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [X] My changes do not add any regressions
- [X] I have tested every sample to ensure everything runs correctly
- [X] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [X] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [X] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

wflohry committed:
Using basic raytracing as template

wflohry committed:
Load sponza

wflohry committed:
Loading sponza model

wflohry committed:
Separating out ray tracing command buffers

wflohry committed:
Add render modes

wflohry committed:
Edit vertex structure

wflohry committed:
Put static objects in buffer with device-exclusive memory

wflohry committed:
Adding vertex and index data to bindings

wflohry committed:
Demonstration of referencing model data from ray tracing shader

wflohry committed:
Implemented texture array and visualization

wflohry committed:
Added fire pit

wflohry committed:
Added ray-traced shadows

wflohry committed:
Adding support for AO

wflohry committed:
Updated model settings

gpx1000 committed:
Fixed a few warnings and spelling errors.

Also fixed the fences problem that causes crash bug from fences on close.

wflohry committed:
Separating out static model loading

wflohry committed:
Separate handling of static / dynamic buffers

wflohry committed:
Added dynamic object and refactored

gpx1000 committed:
Fixed another warning and fixed the fences.

gpx1000 committed:
Merge branch 'feature/raytrace-extended' into warn_removal_and_fence_fix

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp

gpx1000 committed:
Switch to using constants and specializations in order to change values.  Can switch back to uniforms at a later date but the uniforms were proving a render issue for linux.  This should also demonstrate how to use multiple specialization constants in a ray traced environment.

gpx1000 committed:
more warnings removal

wflohry committed:
Merge warn_remove_and_fence_fix

gpx1000 committed:
Merged in warn_removal_and_fence_fix (pull request #1)

Fixed a few warnings and spelling errors.

Approved-by: William Lohry

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

wflohry committed:
Moving instance buffer creation to TLAS construction

wflohry committed:
Adding refraction placeholder

wflohry committed:
Perform BLAS update for dynamic objects and TLAS update every frame.

wflohry committed:
Adding dynamic sinusoidal wave

wflohry committed:
Finished refractive wave

wflohry committed:
Update to refraction

wflohry committed:
Adding separate flame instance

wflohry committed:
Working flame / particle generator

wflohry committed:
Updates to flame and refraction calculation

wflohry committed:
Updated refraction algorithm

wflohry committed:
Added texture

gpx1000 committed:
more warnings removal
move the png to ktx format

gpx1000 committed:
ensure we're pointing to the right place

wflohry committed:
Merge and update

wflohry committed:
Update to ktx commit

gpx1000 committed:
run clang-format over the sample.

wflohry committed:
Added transformations

wflohry committed:
Merged

wflohry committed:
Merged

gpx1000 committed:
Merged in clang-format (pull request #3)

run clang-format over the sample.

Approved-by: William Lohry

wflohry committed:
Adding transformations

wflohry committed:
Implemented automatic  billboard rotation

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

gpx1000 committed:
run a profile pass.
move pts, uvs, normals, indices to be created/initialized once to avoid dynamic memory generation.
remove warnings.
Update copyright.

gpx1000 committed:
Merged in profile (pull request #4)

run a profile pass.

Approved-by: William Lohry

gpx1000 committed:
Update copyright in CMakeLists.txt
Update sample cmake function definition
Update samples list in Samples CMakeLists.txt
Update README.md with first pass definition of the sample.

gpx1000 committed:
Merged in draft_ready (pull request #5)

Update copyright in CMakeLists.txt

Approved-by: William Lohry

wflohry committed:
Added tutorial rough draft

gpx1000 committed:
grammar check.

gpx1000 committed:
grammar check.

gpx1000 committed:
Merged in tutorial (pull request #6)

grammar check.

Approved-by: William Lohry

wflohry committed:
Added fence reset

wflohry committed:
Updating triangle indexing code

wflohry committed:
Avoid dynamic allocation / copy

gpx1000 committed:
update copyright

gpx1000 committed:
Merged in feature/raytrace-extended (pull request #7)

Feature/raytrace extended

gpx1000 committed:
Merge branch 'downstream-master'

gpx1000 committed:
Fixes for Clang-format and copyright headers CI failures.

gpx1000 committed:
add in nonuniformEXT to the texture lookup as imageOffset isn't uniform.

gpx1000 committed:
switch to textureLod from texture to ensure it's ready for mipmaps if copied from sample.

wflohry committed:
Fixing refraction bug

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

gpx1000 committed:
fix typo in README.md

gpx1000 committed:
fix typo in README.md and in CMakeLists.txt

gpx1000 committed:
Merge branch 'master' into master

gpx1000 committed:
Cherry pick from internal bindless branch.

@wlohry thanks for the effort on this!

gpx1000 committed:
fix the README.md

gpx1000 committed:
fix the Copyright headers

cforfang commented on shaders/bindless/bindless.vert:
```diff
@@ -0,0 +1,43 @@
+
+layout(location = 0) out vec4 o_pos;
+layout(location = 1) out vec2 o_uv;
+layout(location = 2) out float o_texture_index;
```
Could this be `out uint` (in VS) and `flat in uint` (in FS) instead of casting to float and back?

cforfang commented on shaders/bindless/cull_address.comp:
```diff
@@ -0,0 +1,98 @@
+{
+	uint             id         = gl_GlobalInvocationID.x;
+	ModelInformation model      = model_buffer.arr[id];
+	bool             is_visible = check_is_visible(global_uniform.proj * global_uniform.view, vec3(model.x, model.y, model.z), model.r);
```
Would it make sense to pre-multiply `global_uniform.proj * global_uniform.view` on CPU instead of in every compute thread?

cforfang commented on shaders/bindless/bindless.vert:
```diff
@@ -0,0 +1,43 @@
+}
+global_uniform;
+
+layout(location = 0) out vec4 o_pos;
```
Unused?

cforfang commented:
These are texture-bleeding artifacts I think? They don't go away with nearest filtering, so probably an issue when generating the textures?
![image](https://user-images.githubusercontent.com/544178/130502851-fe1853bf-c3b2-47a7-b5a7-74cc80168d27.png)
![image](https://user-images.githubusercontent.com/544178/130502904-bcba08e3-c4af-40f2-98ae-37f7cf458ec7.png)

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,925 @@
+	if (gpu.get_features().multiDrawIndirect)
+	{
+		gpu.get_mutable_requested_features().multiDrawIndirect = VK_TRUE;
+		m_supports_mci                                         = true;
```
`m_supports_mci` -> `m_supports_mdi`?

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,925 @@
+	{
+		indirect_flags |= VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT;
+	}
+	indirect_call_buffer = std::make_unique<vkb::core::Buffer>(get_device(), models.size() * sizeof(VkDrawIndexedIndirectCommand), indirect_flags, VMA_MEMORY_USAGE_GPU_ONLY);
```
This buffer is created with `sharingMode VK_SHARING_MODE_EXCLUSIVE`, however from the code it might look like the culling compute dispatches and rendering could run on different queue families. Should this either be `VK_SHARING_MODE_CONCURRENT`, or there needs to be ownership transfer as part of the compute -> graphics handover?

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,925 @@
+
+	update_scene_uniform();
+
+	run_cull();
```
The rendering done in `draw()` above reads the indirect arguments. Is there currently anything synchronizing the work in `run_cull()` which _writes_ the indirect arguments, i.e. preventing a WAR hazard?

(There is a wait after the dispatch in `run_cull()`, preventing RAW hazard. However I also wonder if there should be a pipeline barrier there to guarantee the written data is made available?)

gpx1000 replied:
The synchronization should be corrected explicitly now.  Draw has the device wait idle (that I'd love to take out of the framework), the memory barrier and the fence in culling should be correctly sufficient; please double check.

cforfang replied:
> memory barrier and the fence in culling

Hm, is there supposed to be a memory barrier in `run_gpu_cull()`? Don't think I see it.

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,925 @@
+		image_barrier.oldLayout            = VK_IMAGE_LAYOUT_UNDEFINED;
+		image_barrier.newLayout            = VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL;
+
+		// The semaphore takes care of srcStageMask.
```
Which semaphore does this refer to?

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,925 @@
+		bind(device_address_pipeline, device_address_pipeline_layout, device_address_descriptor_set);
+	}
+
+	vkCmdDispatch(cmd, models.size(), 1, 1);
```
This would effectively dispatch N work-groups consisting of 1 thread, would it? Would it make more sense, performance wise, to dispatch e.g. N/64 workgroups of 64 threads -- so 64-wide warps can be fully filled?

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,925 @@
+
+		auto &queue = device->get_queue_by_flags(VK_QUEUE_GRAPHICS_BIT, 0);
+		queue.submit(texture_cmd, device->request_fence());
+		device->get_fence_pool().wait();
```
Should probably add `device->get_fence_pool().reset();` after this, else the request-pool grows in size for each mesh. This shouldn't _really_ be a problem -- would just result in a bunch of already-waited-on-fenced being redundantly passed to WaitForFences -- however, without the reset, I'm getting occasional DEVICE_LOST on my RTX2060 on Windows.

wflohry committed:
Code updates per PR comments

wflohry committed:
Updating compute workgroup size

wflohry committed:
Adding support for queue families in buffer creation

cforfang commented:
I didn't see any mention of licensing for the assets; should that be mentioned somewhere?

SaschaWillems commented:
> I didn't see any mention of licensing for the assets; should that be mentioned somewhere?

See https://github.com/KhronosGroup/Vulkan-Samples-Assets/pull/13. We already discussed that on a different communication channel, and yes, the license for those assets needs to be added there too.

wflohry committed:
Clang-format

wflohry committed:
Added memory barrier

wflohry committed:
Updated frustum culling plane selection

gpx1000 committed:
fix for various requested comments and culling bug.

gpx1000 commented:
> These are texture-bleeding artifacts I think? They don't go away with nearest filtering, so probably an issue when generating the textures?
> ![image](https://user-images.githubusercontent.com/544178/130502851-fe1853bf-c3b2-47a7-b5a7-74cc80168d27.png)
> ![image](https://user-images.githubusercontent.com/544178/130502904-bcba08e3-c4af-40f2-98ae-37f7cf458ec7.png)

Interesting find.  I'm honestly not sure if it's the model as intended or indeed texture bleeding.  I'll investigate further.

**Update
The problem is in the blender converting to gltf and multiple textures as per requirement.  The conversion resulted in the model not having correctly lined up textures.  We're working on reexporting the model to create a fix.

gpx1000 committed:
fix semaphore errant comment
add fence reset

gpx1000 commented:
> > I didn't see any mention of licensing for the assets; should that be mentioned somewhere?
> 
> See [KhronosGroup/Vulkan-Samples-Assets#13](https://github.com/KhronosGroup/Vulkan-Samples-Assets/pull/13). We already discussed that on a different communication channel, and yes, the license for those assets needs to be added there too.

Attribution should now exist in the README properly

gpx1000 committed:
Merge branch 'master-upstream' into bindless_resources

gpx1000 committed:
add doxygen missing param to framework buffer.h
fix the clang-format issue on framework buffer.cpp and buffer.h

gpx1000 commented:
> **Update
> The problem is in the blender converting to gltf and multiple textures as per requirement. The conversion resulted in the model not having correctly lined up textures. We're working on reexporting the model to create a fix.

This should now be fixed by the latest assets

wflohry committed:
Updates for PR comments

wflohry committed:
Adding memory barriers

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,955 @@
+
+		VkImageMemoryBarrier image_barrier = vkb::initializers::image_memory_barrier();
+		image_barrier.srcAccessMask        = 0;
+		image_barrier.dstAccessMask        = 0;
```
`dstAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT`

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,955 @@
+		image_barrier.oldLayout            = VK_IMAGE_LAYOUT_UNDEFINED;
+		image_barrier.newLayout            = VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL;
+
+		vkCmdPipelineBarrier(texture_cmd.get_handle(), VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, VK_PIPELINE_STAGE_ALL_COMMANDS_BIT, 0, 0, nullptr, 0, nullptr, 1, &image_barrier);
```
`VK_PIPELINE_STAGE_ALL_COMMANDS_BIT` -> `VK_PIPELINE_STAGE_TRANSFER_BIT` -- might as well be specific here.

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,955 @@
+                                                           VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_TRANSFER_DST_BIT,
+                                                           VMA_MEMORY_USAGE_GPU_ONLY,
+                                                           VK_SAMPLE_COUNT_1_BIT,
+                                                           texture.n_mip_maps,
```
Only mip 0 is uploaded below I think -- could one make this `1` and/or assert `texture.n_mip_maps == 1`?

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,955 @@
+		image_barrier.oldLayout            = VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL;
+		image_barrier.newLayout            = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+
+		vkCmdPipelineBarrier(cmd.get_handle(), VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, VK_PIPELINE_STAGE_ALL_COMMANDS_BIT, 0, 0, nullptr, 0, nullptr, 1, &image_barrier);
```
`VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT` -> `VK_PIPELINE_STAGE_TRANSFER_BIT`. Could minimize `VK_PIPELINE_STAGE_ALL_COMMANDS_BIT` to `VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT` as well.

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,955 @@
+		subresource_range.levelCount              = texture.n_mip_maps;
+
+		VkImageMemoryBarrier image_barrier = vkb::initializers::image_memory_barrier();
+		image_barrier.srcAccessMask        = 0;
```
`srcAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT`

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,955 @@
+
+		VkImageMemoryBarrier image_barrier = vkb::initializers::image_memory_barrier();
+		image_barrier.srcAccessMask        = 0;
+		image_barrier.dstAccessMask        = 0;
```
`dstAccessMask = VK_ACCESS_SHADER_READ_BIT`

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,955 @@
+
+	auto &cmd = device->get_command_pool().request_command_buffer();
+	cmd.begin(VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT, VK_NULL_HANDLE);
+	for (auto &&texture : textures)
```
Worth it to do 1x `vkCmdPipelineBarrier` with Nx `VkImageMemoryBarrier`, instead of Nx `vkCmdPipelineBarrier` with 1x `VkImageMemoryBarrier`?

gpx1000 replied:
I don't think there's a significant difference to the two in this particular case but will defer to your judgement if you think it's worth making the sync change.  It should be correct sync in either solution, I'm unaware of the performance or tutorial benefit of switching the two for a small sample like this?

cforfang commented on samples/performance/bindless/bindless.cpp:
```diff
@@ -0,0 +1,955 @@
+	memory_barrier.dstAccessMask         = VK_ACCESS_MEMORY_READ_BIT;
+	memory_barrier.size                  = scene_uniform_buffer->get_size();
+
+	vkCmdPipelineBarrier(memory_cmd.get_handle(), VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, VK_PIPELINE_STAGE_HOST_BIT, 0, 0, VK_NULL_HANDLE, 1, &memory_barrier, 0, VK_NULL_HANDLE);
```
Not sure I understand what this is doing -- add a comment maybe?

gpx1000 replied:
this has been reworked

wflohry committed:
Updating per PR

wflohry committed:
Updating shaders to use defines

gpx1000 committed:
Merge branch 'gpx1000/bindless_resources' into bindless_resources

# Conflicts:
#	samples/performance/bindless/bindless.cpp

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp
#	shaders/khr_ray_tracing_extended/closesthit.rchit

gpx1000 committed:
clang format cleanup

gpx1000 committed:
add missing copyright to new header

gpx1000 committed:
fix attribution in the tutorial.md

gpx1000 committed:
framework:
acceleration structure: (build) fix a bug which only would allow a single triangle to be included in a acceleration structure.
acceleration structure: (build) allow for more than one call to build (useful in animation).  This was caused by recreating the buffer each time rather than reusing it if it already exists.
acceleration structure: (add triangles) allow for developer to specify the data address of the vertex index and transform buffers that are used.  This allows for multi model gltf files to work as in the sponza scene.
acceleration structure: enable buffer retrieval to allow for developer to add in memory barriers.
acceleration structure: provide developer means to wipe the geometries added to the acceleration structure.  This is very important if animating and thus recalling the same structure(s).  NB: a future feature would be to index the geometries and instances to allow for updating existing data individually rather than a requirement to wipe out everything and rebuild everything.
scratch buffer: add a return for the scratch buffer's size to allow for reusing the scratch buffer if it's the same size as previously allocated.

Sample:
As this is the first ray tracing sample to make use of @Sascha William's excellent acceleration structure framework, and we already have the existing long form, I opted to enable a macro to turn on/off using the framework's acceleration structure.  I think this is valuable to end users to show a precise "what the frameork does" behind the scenes.  Future samples should opt to use the framework.

gpx1000 committed:
run clang-format over scratch_buffer.h

gpx1000 committed:
Add in proper update logic for the acceleration_structure in both the framework and the sample.

gpx1000 committed:
Ensure that only updated geometry are used in each update pass and make framework the default AS.

SaschaWillems reviewed (changes requested):
I'm getting two validation layer messages at startup using SDK Version 1.2.189:

```[error] [framework\core\instance.cpp:41] 1165064310 - VUID-VkGraphicsPipelineCreateInfo-layout-00756: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-layout-00756 ] Object 0: handle = 0x16ee604fae0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x45717876 | Shader expects at least 256 descriptors for binding 0.1 but only 225 provided The Vulkan spec states: layout must be consistent with all shaders specified in pStages (https://vulkan.lunarg.com/doc/view/1.2.189.0/windows/1.2-extensions/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-layout-00756)```

```[error] [framework\core\instance.cpp:41] 1028204675 - VUID-VkShaderModuleCreateInfo-pCode-04147: Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-04147 ] Object 0: handle = 0x16ee604fae0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x3d492883 | vkCreateShaderModule(): The SPIR-V Extension (SPV_EXT_physical_storage_buffer) was declared, but none of the requirements were met to use it. The Vulkan spec states: If pCode declares any of the SPIR-V extensions listed in the SPIR-V Environment appendix, one of the corresponding requirements must be satisfied (https://vulkan.lunarg.com/doc/view/1.2.189.0/windows/1.2-extensions/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-04147)```

And also a bunch of validation layer errors at instance destruction with pipelines, pipeline layouts and descriptor sets not being freed.

SaschaWillems commented:
It also looks like the "Freeze culling" option isn't implemented yet. There is a toggle for it. but `m_freeze_cull` is never evaluated.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Small changes for buffer.

wflohry committed:
Updating texture size

wflohry committed:
Adding destructors

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'gpx1000/bindless_resources' into bindless_resources

gpx1000 committed:
Merge branch 'master-upstream' into bindless_resources

gpx1000 commented:
Message From Will regarding the remaining validation error:

> There are two SPV extensions: SPV_EXT_physical_storage_buffer and SPV_KHR_physical_storage_buffer. The current compiler outputs SPV_EXT*, but then the code has to use VK_EXT_BUFFER_DEVICE_ADDRESS_EXTENSION_NAME, but this causes a lot of other problems within the code (e.g. vma allocator).
> Alternatively, if the SPV_KHR_physical_storage_buffer and VK_KHR_BUFFER_DEVICE_ADDRESS is used, then no validation layer messages will be produced. However, I can't figure out how to get the compiler to use the SPV_KHR* extension rather than SPV_EXT* one.
> 

I gave an effort to see if I could get the compiler to output the KHR extension but it doesn't look like that's happening.  Anyone have suggestions?

SaschaWillems commented:
It looks like this behavior has been fixed by this PR: https://github.com/KhronosGroup/glslang/pull/2588

Taking a quick look at my local copy of our repo, it seems we're on an older glslang version.

So I guess we can fix this by updating the glslang submodule on our repo.

gpx1000 committed:
update glslang and switch to using Vulkan 1.2 so that glslang compiler will output SPRIV 1.5 which contains the fix for validation layer.

gpx1000 commented:
Thanks Sascha that does indeed fix it.

gpx1000 committed:
run clang-format.

wflohry committed:
Implementing freeze culling

gpx1000 committed:
Merge branch 'bindless_resources' into gpx1000/bindless_resources

gpx1000 committed:
follow @cforfang's advice on explicitly showing one mipmap assertion to ensure this is clear.

SaschaWillems commented:
The freeze culling option now works as expected 👍 

But it seems like the sample is leaking memory. I've been running for a few minutes to test everything, and noticed memory steadily increasing in VS (same for release/debug) mode with validation disabled:

![image](https://user-images.githubusercontent.com/10283127/134407294-cc0ef04d-29d9-47a8-9ff3-6865721d98ad.png)

I did some memory snapshots, and it seems like the app is leaking several objects on each frame:

![image](https://user-images.githubusercontent.com/10283127/134407382-51c1f224-7372-4233-a341-f7e16746b995.png)

Snapshot 1:
![image](https://user-images.githubusercontent.com/10283127/134407402-109e531d-d98b-4168-8bbd-4f1faaa6499c.png)

Snapshot 4:
![image](https://user-images.githubusercontent.com/10283127/134407427-48abaf71-4335-47b3-b285-7e9f43a99dbc.png)

gpx1000 committed:
fix the memory leak on every run.

LSAN still reports 225403 byte(s) leaked in 791 allocations but those appear to be false positives and/or buried deep in the framework.

gpx1000 commented:
Good catch @SaschaWillems !

gpu_cull was indeed leaking a command buffer.  When testing with LSAN, I'm still seeing 
SUMMARY: LeakSanitizer: 225403 byte(s) leaked in 791 allocation(s)

However, after spending the day tracking all of these down, each of those allocations are either false positives or buried deep in the framework and are incurred no matter what the sample does.  We should do a cleanup pass, but most of the significant leaks found look like false positives (i.e. storing shared_ptr inside a map structure accounts for several).
Do me a favor and retest to confirm that you're not seeing anymore odd memory leaks specific to this sample.  I'll try to find some time to sanitize the framework in a future PR.

gpx1000 committed:
clang-format recent changes.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream' into bindless_resources

gpx1000 committed:
run clang-format

gpx1000 committed:
fix android

gpx1000 committed:
fix android

SaschaWillems commented:
Did a rerun, and as you noticed, most of those leaks are caused by the framework. As this is not the sample's fault and needs fixing in the framework I'm happy with the sample :)

P.S. : Sorry for closing, clicked the wrong button.

gpx1000 committed:
fix android

gary-sweet reviewed (changes requested):
I'm seeing a driver assert about accessing samplers out-of-range.

It would appear that the sample is not respecting maxPerStageDescriptorSamplers (64 in our case).

cforfang commented:
> I'm seeing a driver assert about accessing samplers out-of-range.
> 
> It would appear that the sample is not respecting maxPerStageDescriptorSamplers (64 in our case).

Ah, good find. I suppose the fix is to enable update after bind, to get the guaranteed min limit of maxPerStageDescriptorUpdateAfterBindSamplers = 500000?

gary-sweet commented:
> Ah, good find. I suppose the fix is to enable update after bind, to get the guaranteed min limit of maxPerStageDescriptorUpdateAfterBindSamplers = 500000?

Might be better to just test that maxPerStageDescriptorSamplers is large enough for this sample and flag it as unsupported if it isn't. If the sample doesn't strictly require update-after-bind it seems a shame to require it to workaround a limits issue.

cforfang commented:
> Might be better to just test that maxPerStageDescriptorSamplers is large enough for this sample and flag it as unsupported if it isn't. If the sample doesn't strictly require update-after-bind it seems a shame to require it to workaround a limits issue.

Well, it _is_ a bindless sample :) The only way to get the guaranteed support is to actually use the new feature provided by descriptor_indexing -- so from that angle doing this makes sense to me.

cforfang commented:
Actually I forgot this sample _doesn't_ actually use descriptor_indexing, which is potentially a bit confusing given the references to 'bindless'. But anyway that does maybe mean it makes less sense to add this as a new dependency just to be able to bind more textures/samplers. 

Maybe the way to go is to:
- Check this particular limit -- and any others that may apply. Maybe verify against the Vulkan min specs by using the devsim layer or similar?
- Check against shaderSampledImageArrayDynamicIndexing feature and any others that apply.

One could maybe make an argument that it could make sense to simple depend on descriptor_indexing being available, though. That would guarantee the implementation provides all needed functionality here.

gary-sweet commented:
> One could maybe make an argument that it could make sense to simple depend on descriptor_indexing being available, though. That would guarantee the implementation provides all needed functionality here.

Personally, I'd prefer not to depend on descriptor indexing if possible (since we don't actually support that yet in our driver).

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
address the returns doc.

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge branch 'upstream-master'

wflohry committed:
Adding check of maxPerStageDescriptorSamplers

JoseEmilio-ARM commented:
> > One could maybe make an argument that it could make sense to simple depend on descriptor_indexing being available, though. That would guarantee the implementation provides all needed functionality here.
> 
> Personally, I'd prefer not to depend on descriptor indexing if possible (since we don't actually support that yet in our driver).

I agree with @cforfang (an ARM colleague), for it to be 'bindless' (indexing textures dynamically at runtime) the sample would need to use the descriptor indexing extension.

@gpx1000 I think it would be more appropriate to rename this sample. 'Multi-draw indirect' would be less ambiguous than 'Bindless', what do yo think?

gpx1000 commented:
I'm okay with renaming the sample.  The fact that adding in descriptor indexing would limit the broader support holds a lot of weight with me when it comes to not wanting to add it.  We can rename it for accuracy, maintain the fact that it works in a broad range of devices, and be a very trivial stepping stone one would need to do to get further into bindless territory.
That path makes a lot of sense to me as it does provide a good example of how to do bindless with the benefits of bindless while being more descriptively named due to not implementing descriptor indexing.
What do you guys think?
I'm certainly open to adding in descriptor indexing and getting full bindless as an alternative path but not dropping support isn't a bad thing.

gpx1000 committed:
Merge remote-tracking branch 'downstream/gpx1000/bindless_resources' into gpx1000/bindless_resources

gpx1000 committed:
Merge branch 'master' into gpx1000/bindless_resources

gpx1000 committed:
Merge branch 'gpx1000/bindless_resources' into bindless_resources

gpx1000 committed:
rename bindless to Multi-Draw Indirect

SaschaWillems commented:
Looking good, but I do get one validation layer error with SDK 1.2.189:

```
[error] [framework\core\instance.cpp:41] 1028204675 - VUID-VkShaderModuleCreateInfo-pCode-04147: Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-04147 ] Object 0: handle = 0x1c13c768e80, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x3d492883 | vkCreateShaderModule(): The SPIR-V Extension (SPV_EXT_physical_storage_buffer) was declared, but none of the requirements were met to use it. The Vulkan spec states: If pCode declares any of the SPIR-V extensions listed in the SPIR-V Environment appendix, one of the corresponding requirements must be satisfied (https://vulkan.lunarg.com/doc/view/1.2.189.0/windows/1.2-extensions/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-04147)
```

I think the sample needs to enable the `bufferDeviceAddress` feature of `VkPhysicalDeviceBufferDeviceAddressFeatures` to silence this.

But other than that, the sample works and looks great.

gpx1000 commented:
Hi Sascha,
I think that error is caused by GLSL being an older version.  The one updated in this PR should resolve to SPV_EXT*.  If you can verify that if you're on the same GLSL that this PR has this error goes away it would be appreciated.

SaschaWillems commented:
Yeah, thanks for pointing me in that direction. Switching branches didn't seem to update the submodules, but after a manual update, the false validation errors are gone :)

mklefrancois reviewed (dismissed):
Looking good to me, runs correctly on an NVIDIA GeForce GTX 3060 Ti

gpx1000 committed:
Merge branch 'master' into bindless_resources

SaschaWillems commented:
The sample behaves oddly for me when running a release build: It starts without any errors, but all I can see is the clear color (I have set it to a dark blue to verify this). This happens randomly, but ~ 8 out of 10 times. And it doesn't seem to properly resize when changing the window size:

![image](https://user-images.githubusercontent.com/10283127/142273682-e911bc70-e806-421c-ac38-4aedd72930bf.png)

The UI also doesn't show up, and there are no validation errors.

Sometimes I can get it to draw the scene by wildly resizing, pressing keys and making the window inactive and active again, but most of the time it just looks like that when run as a release build.

I haven't had much time to debug this, but I can imagine that the problem is being caused by uninitialized memory, though I wasn't able to pinpoint this. I am suspecting the `cpu_commands` count at draw time being zero but that is also the case in debug mode.

Btw. I find the naming of `cpu_commands` a bit misleading, as it's also the source of the draw commands when GPU culling is used. So maybe we could rename that one to make it a bit easier to follow?

gpx1000 commented:
Hmm... @SaschaWillems, How'd you get it to behave like that?  I'm not seeing anything that would make an issue happen from Release mode.

SaschaWillems commented:
Just running it in release mode from Visual Studio. But I'm using a Vulkan developer driver, so it may not be the application's fault. If no one else sees this problem I'm fine with approving, cause otherwise the sample looks and works just fine.

SaschaWillems commented:
As discussed on Slack: This seems to be framework related problem that also happens with other samples. Will do a separate issue for that.

The sample itself looks good to me :)

gary-sweet reviewed (commented):
It correctly reports as not supported due to only having 64 samplers, but that's as much as I can test.

gpx1000 committed:
Merge branch 'upstream-master' into bindless_resources

# Conflicts:
#	framework/core/buffer.cpp
#	framework/core/buffer.h

gpx1000 committed:
fix merge issues

gpx1000 committed:
run clang-format

marty-johnson59 commented:
Merging per discussion on Samples call - 3 prior reviews with some minor changes..
